File: crabzilla-vavr-example1/crabzilla-vavr-example1-core/src/main/java/crabzilla/example1/customer/Customer.java
package crabzilla.example1.customer;

import crabzilla.example1.services.SampleInternalService;
import crabzilla.model.Aggregate;
import crabzilla.model.DomainEvent;
import lombok.Value;
import lombok.experimental.Wither;

import java.util.List;

import static crabzilla.example1.customer.CustomerData.*;
import static java.util.Collections.singletonList;
import static org.apache.commons.lang3.Validate.validState;

@Value
@Wither
public class Customer implements Aggregate {

  transient SampleInternalService service;

  CustomerId id;
  String name;
  boolean isActive;
  String reason;

  List<DomainEvent> create(CustomerId id, String name) {
    validState(this.id == null, "customer already created");
    return singletonList(new CustomerCreated(id, name));
  }

  List<DomainEvent> activate(String reason) {
    validState(this.id != null, "unknown customer");
    return singletonList(new CustomerActivated(reason, service.now()));
  }

  List<DomainEvent> deactivate(String reason) {
    validState(this.id != null, "unknown customer");
    return singletonList(new CustomerDeactivated(reason, service.now()));
  }

}

File: crabzilla-kotlin-example1/crabzilla-kotlin-example1-core/src/main/java/crabzilla/example1/customer/Customer.kt
package crabzilla.example1.customer

import crabzilla.example1.SampleInternalService
import crabzilla.example1.eventsOf
import crabzilla.model.Aggregate
import crabzilla.model.DomainEvent
import crabzilla.model.EntityId

data class Customer(val _id: CustomerId? = null,
                    val name: String? = null,
                    val isActive: Boolean? = false,
                    val reason: String? = null,
                    val sampleInternalService: SampleInternalService? = null) : Aggregate {

  override fun getId(): EntityId {
    return _id!!
  }

  internal fun create(id: CustomerId, name: String): List<DomainEvent> {
    require(this._id == null, { "customer already created" })
    return eventsOf(CustomerCreated(id, name))
  }

  internal fun activate(reason: String): List<DomainEvent> {
    return eventsOf(CustomerActivated(reason, sampleInternalService!!.now()))
  }

  internal fun deactivate(reason: String): List<DomainEvent> {
    return eventsOf(CustomerDeactivated(reason, sampleInternalService!!.now()))
  }

}
